enhancements/network/vrf-cni.md

```diff
@@ -0,0 +1,159 @@
+
+Moreover, the multiple networks a pod is connected to may have overlapping CIDRs, and those CIDRs can also overlap with the sdn’s CIDR. To maintain this kind of isolation, they typically use VRFs. Currently this is done by granting the containers the NET_ADMIN capabilities.
+
+### Goals
+
+Being able to assign a secondary interface to a particular vrf, and eventually having more secondary interfaces of the same pod assigned to the same VRF.
+
+### Non-Goals
+
+Assigning host level interfaces to VRFs.
+
+## Proposal
+
+The proposal is to introduce an additional CNI plugin that takes the vrf name as an input, and can be chained to the “main” cni used for the secondary interface, and that has the effect of creating the requested vrf (if it does not exist) and to add the interface to that vrf.
+
+In this way, the network interface returned by any CNI can be isolated in its own VRF, satisfying the above requirements without the need of additional capabilities.
+
+### User Stories [optional]
+
+#### Story 1
+
+As a user, I want to be able to optionally specify a VRF which the secondary nic is attached to.
+
+### Implementation Details/Notes/Constraints
+
+By adding several networks with the same `vrf` name, the CNI will ensure that they are all added to the same vrf, creating it only the first time.
+
+A `NetworkAttachmentDefinition` for MacVlan would look like:
+
+```yaml
+apiVersion: "k8s.cni.cncf.io/v1"
+kind: NetworkAttachmentDefinition
+metadata:
+  name: macvlan-vrf-conf
+spec:
+  config: '{
+              "cniVersion": "0.3.1",
+              "plugins": [
+                {
+                  "type": "macvlan",
+                  "mode": "bridge",
+                  "ipam": {
+                    "type": "host-local",
+                    "subnet": "192.168.1.0/24",
+                    "rangeStart": "192.168.1.200",
+                    "rangeEnd": "192.168.1.216",
+                  }
+                },
+                {
+                  "type": "vrf",
+                  "vrfname": "blue"
+                }
+              ]
+            }'
+```
+
+The same example for a SR-IOV device would look like:
+
+```yaml
+apiVersion: "k8s.cni.cncf.io/v1"
+kind: NetworkAttachmentDefinition
+metadata:
+  name: sriov-vrf-conf
+  annotations:
+    k8s.v1.cni.cncf.io/resourceName: openshift.io/resource_id
+spec:
+  config: '{
+              "cniVersion": "0.3.1",
+              "plugins": [
+                {
+                  "type": "sriov",
+                  "ipam": {
+                    "type": "host-local",
+                    "subnet": "192.168.1.0/24",
+                    "rangeStart": "192.168.1.200",
+                    "rangeEnd": "192.168.1.216",
+                  }
+                },
+                {
+                  "type": "vrf",
```

Comment:
What other configuration parameters are needed? Do we need to have the user specify a route table number?

Reply:
The idea after a bit of discussion is to have only the vrf name as an argument. The cni implementation will pick a valid route table ID making sure it's different from those used by the other vrfs in the pod's namespace.

Reply:
Sure, that makes sense. Nevertheless, I would suggest adding an optional parameter, just in case the user has specific needs.